tests/bsim/bluetooth/mesh/src/test_suspend.c

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include "mesh_test.h"
+
+#include <stddef.h>
+#include <string.h>
+#include <mesh/access.h>
+#include <mesh/net.h>
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+#define LOG_MODULE_NAME test_suspend
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_INF);
+
+#define WAIT_TIME        75 /* seconds */
+#define SUSPEND_START    15 /* seconds */
+#define SUSPEND_DURATION 20 /* seconds */
+#define NUM_PUB			  4 /* Times the transmitter will publish. */
+
+#define TEST_MODEL_ID_1 0x2a2a
+#define TEST_MODEL_ID_2 0x2b2b
+#define TEST_MESSAGE_OP 0x1f
+
+static int model_1_init(struct bt_mesh_model *model);
+static int model_2_init(struct bt_mesh_model *model);
+
+static uint8_t app_key[16] = {0xaa};
+static uint8_t net_key[16] = {0xcc};
+
+static const struct bt_mesh_test_cfg tx_cfg = {
+	.addr = 0x00a0,
+	.dev_key = {0x01},
+};
+static const struct bt_mesh_test_cfg rx_cfg = {
+	.addr = 0x00b0,
+	.dev_key = {0x02},
+};
+
+static struct bt_mesh_prov prov;
+static struct k_sem publish_sem;
+
+static int model_1_update(struct bt_mesh_model *model)
+{
+	model->pub->msg->data[1]++;
+	LOG_INF("Model 1 publishing..., n: %d", model->pub->msg->data[1]);
+	return 0;
+}
+
+static int test_msg_handler(struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx,
+			    struct net_buf_simple *buf)
+{
+	static uint8_t prev_num;
+	uint8_t num = net_buf_simple_pull_u8(buf);
+	uint8_t uptime = k_uptime_get_32() / 1000;
+
+	LOG_INF("Received msg, n: %d", num);
+
+	/* Ensure that payload changes. */
+	ASSERT_TRUE(prev_num != num);
+	prev_num = num;
+
+	/* Ensure that no message is received while Mesh is suspended or disabled. */
+	if (uptime > SUSPEND_START  && uptime < (SUSPEND_START + SUSPEND_DURATION)) {
+		FAIL("Received publication while Mesh is suspended.");
+	}
```

Comment:
You could add additional `else` clauses here to ensure that at least one publication happens outside of this time block, before suspension and after resumption. If because of some other errors, in suspend-resume mechanism, publish never happens, "test_msg_handler()" will never get called and can potentially create false PASS result.

Reply:
With the new use of semaphores, the test will fail if no message is sent/received.